Chm: Add auto-scaled benchmark

zongyao8752593 opened the pull request:
By default, the worker thread count for the CHM benchmark is set to 80.
This patch introduces an auto-scaled benchmark configuration that sets
the thread count to the number of all cpu threads. User can use
`numactl` or `taskset` to limit the cpus to use, and this benchmark will
use the specified cpu number as the worker threads count.

excelle08 commented on packages/chm/ChmBenchmark.cpp:
```diff
@@ -622,6 +624,34 @@ class ChmBenchmark {
+ * Get the cpu core count on single socket
+ */
+int get_cpu_core_count_for_one_socket() {
+    FILE* pipe = popen("lscpu -p=CPU,SOCKET", "r");
```
Please use system calls instead of executing an outside command to determine the CPU core count. We can use [`sched_getaffinity`](https://linux.die.net/man/2/sched_getaffinity).

zongyao8752593 replied:
good idea.

Commit:
Add auto-scaled benchmark for chm

By default, the worker thread count for the CHM benchmark is set to 80.
This patch introduces an auto-scaled benchmark configuration that sets
the thread count to the number of all cpu threads. User can use
`numactl` or `taskset` to limit the cpus to use, and this benchmark will
use the specified cpu number as the worker threads count.

zongyao8752593 commented:
Please review

excelle08 commented on packages/ai_wdl/chm/ChmBenchmark.cpp:
```diff
@@ -37,6 +37,7 @@
 DEFINE_string(distribution_file, "", "Path to the distribution CSV file");
 DEFINE_int32(num_threads, 4, "Number of worker threads per batch");
 DEFINE_int32(num_batch_threads, 2, "Number of parallel batch threads");
+DEFINE_bool(available_cpu_threads, false, "Use all available cpu threads");
```
I think this name for the flag is confusing, a better naming could be `use_all_cpu_threads` or `autoscale`

zongyao8752593 replied:
Done, Thx.

ahmadelyoussef commented:
@zongyao8752593 I collected performance metrics for chm_autoscale_a chm_autoscale_b, the numbers reported in autoscale look like reporting the single thread number. Please multiply by number of threads when reporting.

zongyao8752593 commented:
I add a new PR for the metric change: https://github.com/facebookresearch/DCPerf/pull/371
Please review.

zongyao8752593 commented:
This PR has been merged.
